src/worker.cc

```diff
@@ -616,37 +620,50 @@ void from_json(rapidjson::Document& doc, Options& options) {
 
   // date_time
   auto date_time_type = rapidjson::get_optional<unsigned int>(doc, "/date_time/type");
-  auto date_time_value = rapidjson::get_optional<std::string>(doc, "/date_time/value");
   if (date_time_type && Options::DateTimeType_IsValid(*date_time_type)) {
     // check the type is in bounds
     auto const v = static_cast<Options::DateTimeType>(*date_time_type);
     if (v >= Options::DateTimeType_ARRAYSIZE)
       throw valhalla_exception_t{163};
     options.set_date_time_type(static_cast<Options::DateTimeType>(v));
     // check the value exists for depart at and arrive by
+    auto date_time_value = v != Options::current
+                               ? rapidjson::get_optional<std::string>(doc, "/date_time/value")
+                               : std::string("current");
     if (!date_time_value) {
       if (v == Options::depart_at)
         throw valhalla_exception_t{160};
       else if (v == Options::arrive_by)
         throw valhalla_exception_t{161};
+      else if (v == Options::invariant)
+        throw valhalla_exception_t{165};
     }
-    // check the value is sane for depart at and arrive by
-    if (v != Options::current && !baldr::DateTime::is_iso_valid(*date_time_value))
+    // check the value is sane
+    if (*date_time_value != "current" && !baldr::DateTime::is_iso_valid(*date_time_value))
       throw valhalla_exception_t{162};
-    if (v != Options::current)
-      options.set_date_time(*date_time_value);
-    else
-      options.set_date_time("current");
+    options.set_date_time(*date_time_value);
   } // not specified but you want transit, then we default to current
   else if (options.has_costing() &&
            (options.costing() == multimodal || options.costing() == transit)) {
     options.set_date_time_type(Options::current);
     options.set_date_time("current");
   }
 
+  // failure scenarios with respect to time dependence
+  if (options.has_date_time_type()) {
+    if (options.date_time_type() == Options::arrive_by) {
+      if (options.costing() == multimodal || options.costing() == transit)
+        throw valhalla_exception_t{141};
+      if (options.action() == Options::isochrone)
+        throw valhalla_exception_t{142};
+    } else if (options.date_time_type() == Options::invariant && options.action() != Options::route) {
```

Comment:
we need to resolve this. map matching needs to at least get this date information and use it for costing

Reply:
for now we'll just allow trace_* and in a follow up pr we will actually recost the results of map matching based on whatever time information you have set in the route